File: packages/adapters/nexus-adapter-sdks/nexus-adapter-sdk-ts/src/immutable-record-contract.test.ts
import { createHash } from "node:crypto";
import { describe, expect, it } from "vitest";
import { messageRecord } from "./event.js";
import {
  AdapterInboundRecordSchema,
  CompleteProviderSnapshotSchema,
  completeProviderSnapshot,
} from "./protocol.js";

describe("immutable Record adapter contract", () => {
  it("builds complete provider snapshots and preserves provider provenance", () => {
    const providerObjectJson = '{"id":"456","title":"Fulfillment Runbook","version":{"number":7}}';
    const snapshot = completeProviderSnapshot(providerObjectJson, { page_id: "456" });
    const record = messageRecord({
      platform: "confluence",
      connectionId: "moonsleep-confluence",
      providerAccountRef: "moonsleep.atlassian.net",
      externalRecordId: "ari:cloud:confluence:site-1:page/456",
      sourceRecordType: "page",
      providerVersionRef: "7",
      senderId: "atlassian-user-123",
      containerId: "page-456",
      containerKind: "group",
      content: "Fulfillment runbook snapshot",
      completeProviderSnapshot: snapshot,
    });

    expect(AdapterInboundRecordSchema.parse(record)).toMatchObject({
      routing: { provider_account_ref: "moonsleep.atlassian.net" },
      payload: {
        external_record_id: "ari:cloud:confluence:site-1:page/456",
        source_record_type: "page",
        provider_version_ref: "7",
        payload: {
          provider_object_json: providerObjectJson,
          provider_object_sha256: createHash("sha256").update(providerObjectJson).digest("hex"),
        },
      },
    });
  });

  it("accepts legacy provenance omissions and rejects canonical Nex identity fields", () => {
    const legacy = {
      operation: "record.ingest",
      routing: {
        platform: "gmail",
        connection_id: "casey@moonsleep.co",
        sender_id: "sender@example.com",
        container_id: "sender@example.com",
        container_kind: "direct",
      },
      payload: {
        external_record_id: "gmail:message:abc123",
        timestamp: 1_700_000_000_000,
        content: "Subject: hello",
        content_type: "text",
      },
    };

    expect(() => AdapterInboundRecordSchema.parse(legacy)).not.toThrow();
    for (const field of ["record_id", "canonical_record_id", "identity_sha256", "payload_sha256"]) {
      expect(() =>
        AdapterInboundRecordSchema.parse({
          ...legacy,
          payload: { ...legacy.payload, [field]: "adapter-controlled" },
        }),
      ).toThrow();
    }
  });

  it("rejects partial, decoded, and mismatched complete snapshots", () => {
    const providerObjectJson = '{"id":"gid://shopify/Order/1"}';
    for (const snapshot of [
      { provider_object_json: providerObjectJson },
      {
        provider_object_json: providerObjectJson,
        provider_object_sha256: "0".repeat(64),
      },
      {
        provider_object_json: providerObjectJson,
        provider_object_sha256: createHash("sha256").update(providerObjectJson).digest("hex"),
        provider_object: { id: "gid://shopify/Order/1" },
      },
    ]) {
      expect(() => CompleteProviderSnapshotSchema.parse(snapshot)).toThrow();
    }
  });
});

File: packages/adapters/nexus-adapter-sdks/nexus-adapter-sdk-go/immutable_record_contract_test.go
package nexadapter

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"testing"
)

func TestCompleteProviderSnapshotAndProviderProvenance(t *testing.T) {
	providerObjectJSON := `{"id":"456","title":"Fulfillment Runbook","version":{"number":7}}`
	snapshot, err := CompleteProviderSnapshot(providerObjectJSON, map[string]any{"page_id": "456"})
	if err != nil {
		t.Fatalf("CompleteProviderSnapshot: %v", err)
	}
	digest := sha256.Sum256([]byte(providerObjectJSON))
	if got, want := snapshot["provider_object_sha256"], hex.EncodeToString(digest[:]); got != want {
		t.Fatalf("provider_object_sha256 = %v, want %s", got, want)
	}

	record := MessageRecord(MessageRecordOptions{
		Platform:                 "confluence",
		ConnectionID:             "moonsleep-confluence",
		ProviderAccountRef:       "moonsleep.atlassian.net",
		ExternalRecordID:         "ari:cloud:confluence:site-1:page/456",
		SourceRecordType:         "page",
		ProviderVersionRef:       "7",
		SenderID:                 "atlassian-user-123",
		ContainerID:              "page-456",
		ContainerKind:            "group",
		Content:                  "Fulfillment runbook snapshot",
		CompleteProviderSnapshot: snapshot,
	})
	encoded, err := json.Marshal(record)
	if err != nil {
		t.Fatalf("marshal record: %v", err)
	}
	var roundTrip map[string]any
	if err := json.Unmarshal(encoded, &roundTrip); err != nil {
		t.Fatalf("unmarshal record: %v", err)
	}
	routing := roundTrip["routing"].(map[string]any)
	payload := roundTrip["payload"].(map[string]any)
	if routing["provider_account_ref"] != "moonsleep.atlassian.net" {
		t.Fatalf("provider_account_ref = %v", routing["provider_account_ref"])
	}
	if payload["source_record_type"] != "page" || payload["provider_version_ref"] != "7" {
		t.Fatalf("provider provenance missing: %#v", payload)
	}
	if _, exists := payload["record_id"]; exists {
		t.Fatal("adapter payload unexpectedly contains canonical record_id")
	}
}

func TestCompleteProviderSnapshotRejectsInvalidOrReservedInputs(t *testing.T) {
	for _, test := range []struct {
		name         string
		providerJSON string
		additional   map[string]any
	}{
		{name: "invalid json", providerJSON: "{"},
		{name: "top-level array", providerJSON: "[]"},
		{
			name:         "reserved digest",
			providerJSON: `{"id":"1"}`,
			additional:   map[string]any{"provider_object_sha256": "adapter-controlled"},
		},
	} {
		t.Run(test.name, func(t *testing.T) {
			if _, err := CompleteProviderSnapshot(test.providerJSON, test.additional); err == nil {
				t.Fatal("expected CompleteProviderSnapshot to fail")
			}
		})
	}
}
